app/backend/src/couchers/servicers/gis.py

```diff
@@ -2,70 +2,61 @@
+    )
+
+
+def _query_to_geojson_response(session, query):
+    json_dict = session.query(_build_geojson_query(query)).scalar()
+    return httpbody_pb2.HttpBody(
+        content_type="application/json",
+        # json.dumps escapes non-ascii characters
+        data=json.dumps(json_dict).encode("ascii"),
+    )
+
+
 class GIS(gis_pb2_grpc.GISServicer):
     def GetUsers(self, request, context):
         with session_scope() as session:
-            out = session.execute(
-                """
-            select json_build_object(
-                    'type', 'FeatureCollection',
-                    'features', json_agg(ST_AsGeoJSON(t.*, 'geom', 6)::json)
-                )
-            from (select username, id, geom from users where geom is not null) as t;
-            """
-            )
+            query = session.query(User.username, User.id, User.geom).filter(User.geom != None)
 
-            return httpbody_pb2.HttpBody(
-                content_type="application/json",
-                # json.dumps escapes non-ascii characters
-                data=json.dumps(out.scalar()).encode("ascii"),
-            )
+            return _query_to_geojson_response(session, query)
 
     def GetCommunities(self, request, context):
         with session_scope() as session:
-            out = session.execute(
-                """
-            select json_build_object(
-                    'type', 'FeatureCollection',
-                    'features', json_agg(ST_AsGeoJSON(t.*, 'geom', 6)::json)
-                )
-            from (select * from nodes where geom is not null) as t;
-            """
-            )
-
-            data = json.dumps(out.scalar()).encode("ascii")
+            query = session.query(Node).filter(Node.geom != None)
 
-            return httpbody_pb2.HttpBody(
-                content_type="application/json",
-                data=data,
-            )
+            return _query_to_geojson_response(session, query)
 
     def GetPages(self, request, context):
         with session_scope() as session:
             # TODO: this might get very slow, very quickly
-            out = session.execute(
-                """
-            select json_build_object(
-                    'type', 'FeatureCollection',
-                    'features', json_agg(ST_AsGeoJSON(t.*, 'geom', 6)::json)
-                )
-            from (
-                select page_id as id, geom from page_versions where id in (select max(id) from page_versions group by page_id) and geom is not null
-            ) as t;
-            """
-            )
-
-            data = json.dumps(out.scalar()).encode("ascii")
+            latest_pages = session.query(func.max(PageVersion.id)).group_by(PageVersion.page_id).subquery()
 
-            return httpbody_pb2.HttpBody(
-                content_type="application/json",
-                data=data,
+            query = (
+                session.query(PageVersion.page_id.label("id"), PageVersion.geom)
```

Comment:
Does it matter what we label `page_id` in the subquery? I was trying to drop it but don't know which tests to rerun to see if it breaks...

Reply:
This is very poorly tested currently. But it is used (the `id` goes into a GeoJSON "property") which is used when it's clicked on the map:

https://github.com/Couchers-org/couchers/blob/043ab8bb278f3d733fc7d7cc2fd00be1e60df484/app/frontend/src/features/map/MapPage.tsx#L28-L31

Reply:
That's fine :)